huaweicloud/services/dws/resource_huaweicloud_dws_disaster_recovery.go

```diff
@@ -0,0 +1,543 @@
+				obj := map[string]string{"code": "COMPLETED"}
+				return obj, "COMPLETED", nil
+			}
+			return nil, "ERROR", err
+		}
+
+		status := utils.PathSearch("disaster_recovery.status", respBody, "").(string)
+		unexpectedStatuses := []string{
+			"create_failed", "start_failed", "stop_failed", "abnormal",
+		}
+		if utils.StrSliceContains(unexpectedStatuses, status) {
+			return respBody, "ERROR", fmt.Errorf("unexpect status (%s)", status)
+		}
+
+		if utils.StrSliceContains(targets, status) {
+			return respBody, "COMPLETED", nil
+		}
+		return respBody, "PENDING", nil
+	}
+}
+
+func waitingForActionDisasterRecvery(ctx context.Context, d *schema.ResourceData, client *golangsdk.ServiceClient, target string) error {
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"PENDING"},
+		Target:       []string{"COMPLETED"},
+		Refresh:      disasterRecoveryStatusRefresh(client, d, []string{target}),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		Delay:        30 * time.Second,
+		PollInterval: 30 * time.Second,
+	}
+	_, err := stateConf.WaitForStateContext(ctx)
+	return err
+}
+
+func buildCreateDisasterRecoveryBodyParams(d *schema.ResourceData) map[string]interface{} {
+	return map[string]interface{}{
+		"disaster_recovery": map[string]interface{}{
+			"name":               d.Get("name"),
+			"dr_type":            d.Get("dr_type"),
+			"primary_cluster_id": d.Get("primary_cluster_id"),
+			"standby_cluster_id": d.Get("standby_cluster_id"),
+			"dr_sync_period":     d.Get("dr_sync_period"),
+		},
+	}
+}
+
+func buildUpdateDisasterRecoveryBodyParams(d *schema.ResourceData) map[string]interface{} {
+	return map[string]interface{}{
+		"disaster_recovery": map[string]interface{}{
+			"dr_sync_period": d.Get("dr_sync_period"),
+		},
+	}
+}
+
+func resourceDwsDisasterRecoveryRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		product = "dws"
+	)
+
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return diag.Errorf("error creating DWS client: %s", err)
+	}
+	respBody, err := getDisasterRecovery(client, d)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "DWS disater recovery")
+	}
+	status := utils.PathSearch("disaster_recovery.status", respBody, "").(string)
+	action := d.Get("action")
+	if status == "running" && (action == "" || action == "start") {
+		action = "start"
+	}
+	if status == "running" && action == "switchover" {
+		action = "switchover"
+	}
+	if status == "stopped" {
+		action = "pause"
+	}
+	mErr := multierror.Append(
+		d.Set("region", region),
+		d.Set("name", utils.PathSearch("disaster_recovery.name", respBody, nil)),
+		d.Set("dr_sync_period", utils.PathSearch("disaster_recovery.dr_sync_period", respBody, nil)),
+		d.Set("dr_type", utils.PathSearch("disaster_recovery.dr_type", respBody, nil)),
+		d.Set("status", status),
+		d.Set("action", action),
```

Comment:
The action field does not need to be backfilled.

Reply:
done